File: packages/android/src/main/java/com/kozmos/components/Navbar/Navbar.kt
package com.kozmos.components.navbar

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.Divider
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens
import com.kozmos.tokens.KozmosDimensions

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun KozmosNavbar(
    title: String,
    modifier: Modifier = Modifier,
    navigationIcon: @Composable () -> Unit = {},
    actions: @Composable RowScope.() -> Unit = {}
) {
    KozmosNavbar(
        modifier = modifier,
        logo = { Text(title, style = MaterialTheme.typography.titleMedium) },
        context = { navigationIcon() },
        actions = actions
    )
}

@Composable
fun KozmosNavbar(
    modifier: Modifier = Modifier,
    logo: @Composable RowScope.() -> Unit = {},
    context: @Composable RowScope.() -> Unit = {},
    navigation: @Composable RowScope.() -> Unit = {},
    primaryAction: @Composable RowScope.() -> Unit = {},
    actions: @Composable RowScope.() -> Unit = {},
    utilities: @Composable RowScope.() -> Unit = {},
    account: @Composable RowScope.() -> Unit = {}
) {
    Column(modifier = modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(64.dp)
                .background(KozmosThemeTokens.primitivesColorsBackground0)
                .padding(horizontal = KozmosDimensions.primitivesLayoutSpacing200),
            horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing200),
            verticalAlignment = Alignment.CenterVertically
        ) {
            logo()
            context()
            primaryAction()
            Row(
                modifier = Modifier.weight(1f),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
                content = navigation
            )
            Row(
                horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100),
                verticalAlignment = Alignment.CenterVertically
            ) {
                actions()
                utilities()
                account()
            }
        }
        Divider(color = KozmosThemeTokens.primitivesColorsBackground300)
    }
}

File: packages/ios/Sources/Components/Navbar/Navbar.swift
import SwiftUI

public struct KozmosNavbar: View {
    let logo: AnyView?
    let context: AnyView?
    let navigation: AnyView?
    let primaryAction: AnyView?
    let actions: AnyView?
    let utilities: AnyView?
    let account: AnyView?

    public init<L: View, T: View>(
        title: String,
        @ViewBuilder leading: () -> L,
        @ViewBuilder trailing: () -> T
    ) {
        self.logo = AnyView(Text(title).font(KozmosTypography.headline))
        self.context = AnyView(leading())
        self.navigation = nil
        self.primaryAction = nil
        self.actions = AnyView(trailing())
        self.utilities = nil
        self.account = nil
    }

    public init(title: String) {
        self.logo = AnyView(Text(title).font(KozmosTypography.headline))
        self.context = nil
        self.navigation = nil
        self.primaryAction = nil
        self.actions = nil
        self.utilities = nil
        self.account = nil
    }

    public init<
        Logo: View,
        Context: View,
        Navigation: View,
        PrimaryAction: View,
        Actions: View,
        Utilities: View,
        Account: View
    >(
        @ViewBuilder logo: () -> Logo,
        @ViewBuilder context: () -> Context,
        @ViewBuilder navigation: () -> Navigation,
        @ViewBuilder primaryAction: () -> PrimaryAction,
        @ViewBuilder actions: () -> Actions,
        @ViewBuilder utilities: () -> Utilities,
        @ViewBuilder account: () -> Account
    ) {
        self.logo = AnyView(logo())
        self.context = AnyView(context())
        self.navigation = AnyView(navigation())
        self.primaryAction = AnyView(primaryAction())
        self.actions = AnyView(actions())
        self.utilities = AnyView(utilities())
        self.account = AnyView(account())
    }

    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing200) {
            if let logo = logo {
                logo
                    .frame(alignment: .leading)
            }
            if let context = context {
                context
                    .frame(alignment: .leading)
            }
            if let primaryAction = primaryAction {
                primaryAction
                    .frame(alignment: .leading)
            }
            if let navigation = navigation {
                navigation
                    .frame(maxWidth: .infinity, alignment: .center)
            } else {
                Spacer(minLength: KozmosDimensions.primitivesLayoutSpacing200)
            }
            HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                if let actions = actions {
                    actions
                }
                if let utilities = utilities {
                    utilities
                }
                if let account = account {
                    account
                }
            }
            .frame(alignment: .trailing)
        }
        .frame(minHeight: 64)
        .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing200)
        .background(KozmosColors.primitivesColorsBackground0)
        .overlay(
            Rectangle()
                .frame(height: 1)
                .foregroundColor(KozmosColors.primitivesColorsBackground300),
            alignment: .bottom
        )
    }
}
